app/handlers/handlers_services.py

```diff
@@ -0,0 +1,94 @@
+import kopf
+import kubernetes
+
+from app.handlers import success
+
+
+def k8s_get_twingate_resource(
+    namespace: str, name: str, kapi: kubernetes.client.CustomObjectsApi | None = None
+) -> dict | None:
+    kapi = kapi or kubernetes.client.CustomObjectsApi()
+    try:
+        return kapi.get_namespaced_custom_object(
+            "twingate.com", "v1beta", namespace, "twingateresources", name
+        )
+    except kubernetes.client.exceptions.ApiException as ex:
+        if ex.status == 404:
+            return None
+        raise
+
+
+def service_to_twingate_resource(service_body, namespace) -> dict:
+    meta = service_body.metadata
+    spec = service_body.spec
+    service_name = service_body.meta.name
+    resource_object_name = f"{service_name}-resource"
+
+    result: dict = {
+        "apiVersion": "twingate.com/v1beta",
+        "kind": "TwingateResource",
+        "metadata": {
+            "name": resource_object_name,
+        },
+        "spec": {
+            "name": resource_object_name,
+            "address": f"{service_name}.{namespace}.svc.cluster.local",
+        },
+    }
+
+    for key in ["alias", "isBrowserShortcutEnabled", "securityPolicyId", "isVisible"]:
+        if value := meta.annotations.get(f"twingate.com/expose-{key}"):
+            result["spec"][key] = value
+
+    if service_ports := spec.get("ports", []):
+        protocols: dict = {
+            "allowIcmp": False,
+            "tcp": {"policy": "RESTRICTED", "ports": []},
+            "udp": {"policy": "RESTRICTED", "ports": []},
+        }
+        for port_obj in service_ports:
+            port = port_obj["port"]
+            if port_obj["protocol"] == "TCP":
+                protocols["tcp"]["ports"].append({"start": port, "end": port})
+            elif port_obj["protocol"] == "UDP":
+                protocols["udp"]["ports"].append({"start": port, "end": port})
+
+        result["spec"]["protocols"] = protocols
+
+    return result
+
+
+@kopf.on.resume("services", annotations={"twingate.com/expose": "true"})
+@kopf.on.create("service", annotations={"twingate.com/expose": "true"})
+@kopf.on.update("service", annotations={"twingate.com/expose": "true"})
+def twingate_service_create(body, spec, namespace, meta, logger, **_):
+    logger.info("twingate_service_create: %s", spec)
+    service_name = body.meta.name
+    resource_object_name = f"{service_name}-resource"
+
+    resource_subobject = service_to_twingate_resource(body, namespace)
```

Comment:
The `resource_object_name` is also created in `service_to_twingate_resource()` which might lead to inconsistency. I would either pass this `resource_object_name` to `service_to_twingate_resource()`, or having `service_to_twingate_resource()` method return this value.